src/gpgmm/common/SlabMemoryAllocator.cpp

```diff
@@ -371,32 +385,52 @@ namespace gpgmm {
         SlabBlock* blockInSlab = static_cast<SlabBlock*>(subAllocation->GetBlock());
         ASSERT(blockInSlab != nullptr);
 
-        Slab* slab = blockInSlab->pSlab;
-        ASSERT(slab != nullptr);
+        Slab* pSlab = *(blockInSlab->ppSlab);
+        ASSERT(pSlab != nullptr);
+
+        SlabCache* pCache = GetOrCreateCache(pSlab->Allocation.GetSize());
+        ASSERT(pCache != nullptr);
 
-        // Splice the slab from the full-list to free-list.
-        if (slab->IsFull()) {
-            SlabCache* cache = GetOrCreateCache(slab->Allocation->GetSize());
-            cache->FullList.remove(slab);
-            cache->FreeList.push_front(slab);
+        // Move the slab from the full-list to free-list.
+        if (pSlab->IsFull()) {
+            pSlab = MoveSlabInCache(pSlab, pCache, /*pSrcList*/ &pCache->FullList,
+                                    /*pDstList*/ &pCache->FreeList);
         }
 
         mInfo.UsedBlockCount--;
         mInfo.UsedBlockUsage -= blockInSlab->Size;
 
-        slab->Allocator.DeallocateBlock(blockInSlab);
-        slab->AllocatedBlocks--;
+        pSlab->Allocator.DeallocateBlock(blockInSlab);
+        pSlab->UsedBlocksPerSlab--;
 
         MemoryBase* slabMemory = subAllocation->GetMemory();
         ASSERT(slabMemory != nullptr);
 
         slabMemory->RemoveSubAllocationRef();
 
-        if (slab->IsEmpty()) {
-            mMemoryAllocator->DeallocateMemory(std::move(slab->Allocation));
+        if (pSlab->IsEmpty()) {
+            mMemoryAllocator->DeallocateMemory(
+                std::make_unique<MemoryAllocation>(pSlab->Allocation));
+            pSlab->Allocation = {};  // Invalidate it
         }
     }
 
+    Slab* SlabMemoryAllocator::MoveSlabInCache(Slab* pSlab,
+                                               SlabCache* pCache,
+                                               StableList<Slab>* pSrcList,
+                                               StableList<Slab>* pDstList) {
+        const uint64_t srcIndex = pSlab->IndexInList;
+        pSlab->IndexInList = pDstList->size();
+        pDstList->push_back(*pSlab);  // copy
+        pSrcList->erase(srcIndex);
```

Comment:
Erase causes fragmentation if the erased slab isn't on the end of the list, and it appears to me that at least in the case of `DeallocateMemory()`, calls to `MoveSlabInCache` will be fairly random in terms of the position in the StableList that is being deleted.

I'm wondering why you're maintaining both full/free lists of slabs instead of just using one list and maintaining a list of freed locations within it. You could then reuse the memory and never worry about fragmentation.

Reply:
Yes but that list of freed locations would need to be sorted to stay contiguous in-memory.

The use of stable list is a trade off between contiguous access and space (not both). It will fragment if slabs, of different lifetimes are not used together; otherwise, erase will reclaim from the end, just not always immediately, since order can be random. The use of different lists for the cache is unrelated - that's for speeding up the request-time to GPU memory.

Reply:
Fragmentation seems concerning in general to me, but I'll take your word for it. I'm just imagining someone running an app for a long period of time and things getting very, very fragmented.

Reply:
The number of slabs in the slab cache actually decreases over-time. This is because GPU heaps are re-sized, larger ones replace smaller ones. Or if the heap size is known ahead-of-time, you basically have/coverage to 1 slab. If it didn't do this, I would too be more concerned.

But for the CPU caches, where objects grow unbounded, you are absolutely correct - stable list is a bad idea.